### PR TITLE
feat(firestore): EntityFilter for AND/OR queries

### PR DESCRIPTION
**Pull request by bhshkh**
Adding new method 'WhereEntity' to Query which can be used to construct composite queries.
Adding the following new structs to be used with FilterEntity

- AndFilter
- OrFilter
- PropertyFilter
- PropertyPathFilter

### CODE REVIEW

File: firestore/query.go [hunk 1]
```diff
@@ -132,7 +132,17 @@ func (q Query) Where(path, op string, value interface{}) Query {
 // The op argument must be one of "==", "!=", "<", "<=", ">", ">=",
 // "array-contains", "array-contains-any", "in" or "not-in".
 func (q Query) WherePath(fp FieldPath, op string, value interface{}) Query {
-	proto, err := filter{fp, op, value}.toProto()
+	return q.WhereEntity(PropertyPathFilter{fp, op, value})
```
Comment (enocom) [inline firestore/query.go:135]: nit: Naming the fields in the `PropertyPathFilter` would be nice.

File: firestore/query.go [hunk 2]
```diff
@@ -643,21 +655,143 @@ func (q Query) compareFunc() func(d1, d2 *DocumentSnapshot) (int, error) {
...
-type filter struct {
-	fieldPath FieldPath
-	op        string
-	value     interface{}
+// EntityFilter represents a firestore filter.
+type EntityFilter interface {
+	toProto() (*pb.StructuredQuery_Filter, error)
+}
+
+// CompositeFilter represents firestore composite filters.
```
Comment (enocom) [inline firestore/query.go:663]: Firestore instead of firestore?

**Review by enocom — APPROVED** [review]

File: firestore/integration_test.go [hunk 3]
```diff
@@ -646,6 +703,153 @@ func TestIntegration_WriteBatch(t *testing.T) {
...
+		orderBy bool // Some query types do not allow ordering.
+	}{
+		{
+			desc: "height == 5",
+			q: q.WhereEntity(PropertyFilter{
+				Path:     "height",
+				Operator: "==",
+				Value:    5,
+			}),
+			want:    wants[4:5],
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/integration_test.go:750]: wants is an array. So, wants[4] which is a single element cannot be used

File: firestore/query.go (hunk 2 above)
Comment (bhshkh) [reply firestore/query.go]: Done

File: firestore/query.go (hunk 1 above)
Comment (bhshkh) [reply firestore/query.go]: Done

File: firestore/integration_test.go [hunk 4]
```diff
@@ -646,6 +703,153 @@ func TestIntegration_WriteBatch(t *testing.T) {
...
+							Value:    8,
+						},
+					},
+				},
+			),
+			want:    wants[5:7],
+			orderBy: true,
+		},
+		{
+			desc: "",
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/integration_test.go]: Done

File: firestore/integration_test.go [hunk 5]
```diff
@@ -103,18 +108,70 @@ func initIntegrationTest() {
...
+
+	createIndexes(ctx, wantDBPath)
+
 	refDoc := iColl.NewDoc()
 	integrationTestMap["ref"] = refDoc
 	wantIntegrationTestMap["ref"] = refDoc
 	integrationTestStruct.Ref = refDoc
 }
 
+func createIndexes(ctx context.Context, dbPath string) {
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/integration_test.go:130]: Done

**Review by codyoss — APPROVED** [review]
Stamping for approval, deferring to @enocom review.

**Pull request merged**